src/useDynamicAttribute/index.tsx

```diff
@@ -0,0 +1,38 @@
+import React, { useState } from 'react'
+
+type Enrichment<T, TKey extends string, TSet extends string, TValue> =
+  & Record<TKey, TValue>
+  & Record<TSet, (value: TValue, data: (T & Enrichment<T, TKey, TSet, TValue>)[]) => void>
+
+export type Enriched<T, TKey extends string, TSet extends string, TValue> = T & Enrichment<T, TKey, TSet, TValue>
+
+export function useDynamicAttribute<
+  T extends {},
+  TKey extends string,
+  TSet extends string,
+  TValue
+> (items: T[], key: TKey, set: TSet, initVal: (((item: T) => TValue)) | TValue): (T & Enrichment<T, TKey, TSet, TValue>)[] {
+  type Data = (T & Enrichment<T, TKey, TSet, TValue>)
+
+  function getInitialState (item: T) {
+    if (typeof initVal !== 'function') {
+      return initVal
+    }
+    return (initVal as ((item: T) => TValue))(item)
+  }
+
+  const [data, setData] = useState(items = items.map((dataItem, i) => {
```

Comment:
Items is unused - Also, its not very readable if you have a function in useState - rather create a descriptive const and use the const